dev/buildtool/spin_commands.py

```diff
@@ -171,6 +165,34 @@ def _do_repository(self, repository):
       logging.info('Built spin version {} written to output file {}'
                    .format(self.__build_version, built_version_file))
 
+  def __determine_internal_version(self, repository):
+    # Note: spin CLI is coupled to the Gate major and minor version.
+    # Gate is a routing server, so features and breaking changes in Gate
+    # must be reflected in spin since it is a client.
+    git_dir = repository.git_dir
+    git = self.__scm.git
+    match = re.match(r'(\d+)\.(\d+)\.(\d+)-\d+', self.__gate_version)
+    gate_major = match.group(1)
+    gate_min = match.group(2)
+    if match is None:
+      raise_and_log_error(
+          ConfigError('gate version {version} is not X.Y.Z-<buildnum>'
+                      .format(version=self.__gate_version)))
+
+    tag_matcher = re.compile(r'version-{maj}.{min}.(\d+)'
+                             .format(maj=gate_major, min=gate_min))
+    tags = git.fetch_tags(git_dir)
+    tag_matches = [tag_matcher.match(t) for t in tags if tag_matcher.match(t)]
+    if tag_matches:
+      patch_versions = [int(m.group(1)) for m in tag_matches]
+      max_patch = max(patch_versions)
+      patch = str(max_patch + 1)
+      self.__internal_version = '{major}.{minor}.{patch}'.format(
```

Comment:
why set this rather than return it?
You only use it in the one place calling this function.

Reply:
Done.